src/main/java/jenkins/plugins/slack/pipeline/SlackUploadFileRunner.java

```diff
@@ -0,0 +1,119 @@
+package jenkins.plugins.slack.pipeline;
+
+import hudson.FilePath;
+import hudson.model.TaskListener;
+import hudson.util.DirScanner;
+import hudson.util.FileVisitor;
+import java.io.File;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.plugins.slack.HttpClient;
+import jenkins.security.MasterToSlaveCallable;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.ResponseHandler;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.client.methods.RequestBuilder;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.mime.HttpMultipartMode;
+import org.apache.http.entity.mime.MultipartEntityBuilder;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.util.EntityUtils;
+import org.json.JSONObject;
+
+class SlackUploadFileRunner extends MasterToSlaveCallable<Boolean, Throwable> implements Serializable {
+
+    private static final long serialVersionUID = 1L;
+    private static final String API_URL = "https://slack.com/api/files.upload";
+    private static final Logger logger = Logger.getLogger(SlackUploadFileRunner.class.getName());
+
+    private final FilePath filePath;
+    private String fileToUploadPath;
+
+    private final String channels;
+
+    private final String token;
+
+    private final TaskListener listener;
+    private final String initialComment;
+
+    SlackUploadFileRunner(TaskListener listener, SlackFileRequest slackFileRequest) {
+        this.listener = listener;
+        this.filePath = slackFileRequest.getFilePath();
+            this.fileToUploadPath = slackFileRequest.getFileToUploadPath();
+        this.channels = slackFileRequest.getChannels();
+        this.initialComment = slackFileRequest.getInitialComment();
+        this.token = slackFileRequest.getToken();
+    }
+
+    @Override
+    public Boolean call() throws Throwable {
+        listener.getLogger().println(String.format("Using dirname=%s and includeMask=%s", filePath.getRemote(), fileToUploadPath));
+
+        final List<File> files = new ArrayList<>();
+        new DirScanner.Glob(fileToUploadPath, null).scan(new File(filePath.getRemote()), new FileVisitor() {
+            @Override
+            public void visit(File file, String relativePath) {
+                if (file.isFile()) {
+                    listener.getLogger().println("Adding file (only first one will be uploaded) " + file.getAbsolutePath());
+
+                    files.add(file);
+                }
+            }
+        });
+
+        if (files.isEmpty()) {
+            listener.getLogger().println("No files found for mask=" + this.filePath);
```

Comment:
Hi @casz @oleg-nenashev 
My first time writing some code for remoting, functionally it seems to work fine, with one issue.
The logs (i.e. this statement, no files found) aren't appearing in the build log when running on an agent, (master works fine)
If you have time to take a glance of this it would be appreciated to see where I have gone wrong

I've checked the 'documentation' https://wiki.jenkins.io/display/JENKINS/Making+your+plugin+behave+in+distributed+Jenkins
https://jenkins.io/doc/developer/architecture/remoting/
but it's marked as TBD for this part

Reply:
Was accidentally trying to pass the `Jenkins` instance through... It worked sometimes =/